lib/lambda.js

```diff
@@ -2,11 +2,275 @@
+  const faas = {
+    coldstart: isColdStart,
+    execution: context.awsRequestId,
+    trigger: {
+      type: 'http'
+    }
+  }
+  return faas
+}
+
+function setApiGatewayData (transaction, event, context, isColdStart) {
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.request_id = event && event.requestContext && event.requestContext.requestId
+  faas.trigger.type = 'http'
+  transaction.setFaas(faas)
+
+  transaction.type = 'http'
+  let httpMethod = event && event.requestContext && event.requestContext.httpMethod
+  if (!httpMethod) {
+    httpMethod = event && event.requestContext && event.requestContext.http && event.requestContext.http.method
+  }
+  let path = event && event.requestContext && event.requestContext.resourcePath
+  if (!path) {
+    path = event && event.requestContext && event.requestContext.http && event.requestContext.http.path
+  }
+
+  const stage = event && event.requestContext && event.requestContext.stage
+  const apiId = event && event.requestContext && event.requestContext.apiId
+  transaction.name = httpMethod + ' ' + context.functionName
+  const version = (event && event.version) ? event.version : '1.0'
+  const serviceContext = {
+    origin: {
+      name: `${httpMethod} ${path}/${stage}`,
+      id: apiId,
+      version: version
+    }
+  }
+  transaction.setServiceContext(serviceContext)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws',
+      service: {
+        name: 'api gateway'
+      },
+      account: {
+        id: event && event.requestContext.accountId
+      }
+    }
+  }
+  transaction.setCloudContext(cloudContext)
+}
+
+function isApiGatewayEvent (event) {
+  return !!(event && event.requestContext && event.requestContext.requestId)
+}
+
+function setSqsData (agent, transaction, event, context, isColdStart) {
+  const record = event && event.Records && event.Records[0]
+  record.eventSourceARN = record.eventSourceARN ? record.eventSourceARN : ''
+
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.request_id = record.messageId
+  faas.trigger.type = 'pubsub'
+  transaction.setFaas(faas)
+
+  const arnParts = record.eventSourceARN.split(':')
+  const queueName = arnParts[5]
+  const accountId = arnParts[4]
+
+  transaction.name = `RECEIVE ${queueName}`
+  transaction.type = 'messaging'
+
+  const serviceContext = {
+    origin: {
+      name: queueName,
+      id: record.eventSourceARN
+    }
+  }
+  transaction.setServiceContext(serviceContext)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws',
+      region: record.awsRegion,
+      service: {
+        name: 'sqs'
+      },
+      account: {
+        id: accountId
+      }
+    }
+  }
+  transaction.setCloudContext(cloudContext)
+
+  let age
+  if (record.attributes && record.attributes.SentTimestamp) {
+    age = Date.now() - record.attributes.SentTimestamp
+  }
+  const messageContext = {
+    queue: record.eventSourceARN,
+    age: age,
```

Comment:
Is the https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-aws-lambda.md#sqs--sns wrong here? The APM Server transaction.json spec says these should be `context.message.queue.name` and `context.message.age.ms`.

Reply:
Honoring APM Server spec for now since it's deployed the the world.  Changed in 61ef4f8d0025763b8e8429e29746b1c385bd2d13